File: SingleNumber/SingleNumberIII/SingleNumberIII.cpp
#include <vector>

using namespace std;

class Solution {
public:
    vector<int> singleNumber(vector<int>& nums) {
        int diff = accumulate(nums.begin(), nums.end(), 0, bit_xor<int>());
        diff &= -diff;
        vector<int> rets = {0, 0};
        for(auto num : nums) {
        	if( (num & diff) == 0) {
        		rets[0]^ = num;
        	}
        	else {
        		rets[1]^ = num;
        	}
        }
        return rets;
    }
};

File: SingleNumber/SingleNumberIII/SingleNumberIII.py
class Solution(object):
    def singleNumber(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        tmp = 0
        a = b = 0
        for i in nums:
            tmp ^= i
        l = tmp & (-tmp) # find the fitst no zero bit
        for i in nums:
            if i & l:
                a ^= i
            else:
                b ^= i
        return [a, b]